src/lib/Microsoft.Health.Fhir.Ingest.Template/CalculatedFunction/MeasurementExtractor.cs

```diff
@@ -0,0 +1,166 @@
+                    }
+
+                    throw new InvalidOperationException($"Unable to extract required value for [{name}] using {string.Join(",", expressions.Select(e => e.Value).ToArray())}");
+                }
+            }
+            else if (isRequired)
+            {
+                throw new InvalidOperationException($"An expression must be set for [{name}]");
+            }
+
+            return default;
+        }
+
+        protected static bool IsExpressionDefined(params Expression[] expressions)
+        {
+            EnsureArg.IsNotNull(expressions, nameof(expressions));
+
+            return expressions.Any(ex => ex != null && !string.IsNullOrWhiteSpace(ex.Value));
+        }
+
+        protected virtual DateTime? GetTimestamp(JToken token) => EvalExpression<DateTime?>(token, nameof(Template.TimestampExpression), true, Template.TimestampExpression);
+
+        protected virtual string GetDeviceId(JToken token) => EvalExpression<string>(token, nameof(Template.DeviceIdExpression), true, Template.DeviceIdExpression);
+
+        protected virtual string GetPatientId(JToken token) => EvalExpression<string>(token, nameof(Template.PatientIdExpression), false, Template.PatientIdExpression);
+
+        protected virtual string GetEncounterId(JToken token) => EvalExpression<string>(token, nameof(Template.EncounterIdExpression), false, Template.EncounterIdExpression);
+
+        protected virtual string GetCorrelationId(JToken token) => EvalExpression<string>(token, nameof(Template.CorrelationIdExpression), true, Template.CorrelationIdExpression);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        protected virtual IEnumerable<JToken> MatchTypeTokens(JObject token)
+        {
+            EnsureArg.IsNotNull(token, nameof(token));
+            var evaluator = ExpressionEvaluatorFactory.Create(Template.TypeMatchExpression);
+
+            foreach (var extractedToken in evaluator.SelectTokens(token))
+            {
+                // Add the extracted data as an element of the original data.
+                // This allows subsequent expressions access to data from the original event data
+
+                var tokenClone = token.DeepClone() as JObject;
+                tokenClone.Add(MatchedToken, extractedToken);
+                yield return tokenClone;
+            }
+        }
+
+        private Measurement CreateMeasurementFromToken(JToken token)
+        {
+            // Current assumption is that the expressions should match a single element and will error otherwise.
+
+            string deviceId = GetDeviceId(token);
+            DateTime? timestamp = GetTimestamp(token);
+
+            string correlationId = null;
+            if (IsExpressionDefined(Template.CorrelationIdExpression))
+            {
+                correlationId = GetCorrelationId(token);
+            }
```

Comment:
I see below that we simple call GetPatientId(token) and GetEncounterId(token) without doing this "IsExpressionDefined" check - Is there a reason we need to treat the correlationId differently?

Reply:
Interesting question. This logic is in the existing [JsonPathContentTemplate](https://github.com/microsoft/iomt-fhir/blob/master/src/lib/Microsoft.Health.Fhir.Ingest.Template/JsonPathContentTemplate.cs#L115). So I kept that logic here. 

My assumption was that `CorrelationId` isn't required, but if its specified then it must return a value.

Reply:
That is correct.  The correlation id support came later.  When I added it I thought about doing the same check for PatientId and EncounterId but didn't because it would change current behavior.

We can revisit if we would like.  Same question was brought up in the PR that introduced the correlation id support; https://github.com/microsoft/iomt-fhir/pull/47/files